packages/ui-extensions/src/surfaces/checkout/api/shared.ts

```diff
@@ -148,3 +148,34 @@ export interface ShippingAddress extends MailingAddress {
    */
   oneTimeUse?: boolean;
 }
+
+type MaybePromise<T> = T extends Promise<any> ? T : T | Promise<T>;
+type Subscriber<T> = (value: T) => unknown;
+type RemoteSubscribeResult<T> = [() => void, T];
+
+/**
+ * A remote-subscribable object managed on the main thread.
+ *
+ * Example: Checkout uses this to manage the state of an object and
+ * communicate state changes to extensions running in a sandboxed web worker.
+ */
+export interface RemoteSubscribable<T> {
+  readonly initial: T;
+  subscribe(subscriber: Subscriber<T>): MaybePromise<RemoteSubscribeResult<T>>;
+}
+
+/**
+ * A remote-subscribable object exposed to a sandboxed web worker.
+ *
+ * Example: extensions use this to get updates about an object
+ * managed by Checkout on the main thread.
+ *
+ * This interface is compatible with Preact's ReadonlySignal:
+ * https://github.com/preactjs/signals/blob/a023a132a81bd4ba4a0bebb8cbbeffbd8c8bbafc/packages/core/src/index.ts#L700-L709
+ */
+export interface StatefulRemoteSubscribable<T> {
+  readonly value: T;
+  readonly current: T;
+  subscribe(subscriber: Subscriber<T>): () => void;
+  destroy(): Promise<void>;
```

Comment:
I know this is likely more for backward compatibility, but doesn't the subscribe method return a teardown that would achieve the same thing?